runners/pvc_autoresizer.go

```diff
@@ -0,0 +1,255 @@
+
+	return &PVCAutoresizer{
+		metricsClient: mc,
+		interval:      interval,
+		recorder:      recorder,
+	}
+}
+
+// InjectClient implements inject.Client
+func (w *PVCAutoresizer) InjectClient(c client.Client) error {
+	w.client = c
+	return nil
+}
+
+// InjectLogger implements inject.Logger
+func (w *PVCAutoresizer) InjectLogger(log logr.Logger) error {
+	w.log = log
+	return nil
+}
+
+// PVCAutoresizer is a runner which resize PVC capacity automatically
+type PVCAutoresizer struct {
+	client        client.Client
+	metricsClient MetricsClient
+	interval      time.Duration
+	log           logr.Logger
+	recorder      record.EventRecorder
+}
+
+// Start implements manager.Runnable
+func (w *PVCAutoresizer) Start(ch <-chan struct{}) error {
+	ticker := time.NewTicker(w.interval)
+	ctx := context.Background()
+
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ch:
+			return nil
+		case <-ticker.C:
+			err := w.notifyPVCEvent(ctx)
+			if err != nil {
+				w.log.Error(err, "failed to notifyPVCEvent")
+				return err
+			}
+		}
+	}
+}
+
+func isTargetPVC(pvc *corev1.PersistentVolumeClaim) bool {
+	if pvc.Spec.Resources.Limits.Storage().IsZero() {
+		return false
+	}
+	if pvc.Spec.VolumeMode != nil && *pvc.Spec.VolumeMode != corev1.PersistentVolumeFilesystem {
+		return false
+	}
+	if pvc.Status.Phase != corev1.ClaimBound {
+		return false
+	}
+	return true
+}
+
+func (w *PVCAutoresizer) getStorageClassList(ctx context.Context) (*storagev1.StorageClassList, error) {
+	var scs storagev1.StorageClassList
+	err := w.client.List(ctx, &scs, client.MatchingFields(map[string]string{resizeEnableIndexKey: "true"}))
+	if err != nil {
+		return nil, err
+	}
+	return &scs, nil
+}
+
+func (w *PVCAutoresizer) notifyPVCEvent(ctx context.Context) error {
+	scs, err := w.getStorageClassList(ctx)
+	if err != nil {
+		return err
+	}
+
+	vsMap, err := w.metricsClient.GetMetrics(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, sc := range scs.Items {
+		var pvcs corev1.PersistentVolumeClaimList
+		err = w.client.List(ctx, &pvcs, client.MatchingFields(map[string]string{storageClassNameIndexKey: sc.Name}))
+		if err != nil {
+			return err
+		}
+		for _, pvc := range pvcs.Items {
+			if !isTargetPVC(&pvc) {
+				continue
+			}
+			namespacedName := types.NamespacedName{
+				Namespace: pvc.Namespace,
+				Name:      pvc.Name,
+			}
+			if _, ok := vsMap[namespacedName]; !ok {
+				continue
+			}
+			err = w.resize(ctx, &pvc, vsMap[namespacedName])
+			if err != nil {
+				// TODO
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+func (w *PVCAutoresizer) resize(ctx context.Context, pvc *corev1.PersistentVolumeClaim, vs *VolumeStats) error {
+	log := w.log.WithName("resize").WithValues("namespace", pvc.Namespace, "name", pvc.Name)
+
+	threshold, err := convertSizeInBytes(pvc.Annotations[ResizeThresholdAnnotation], vs.CapacityBytes, DefaultThreshold)
+	if err != nil {
+		return err
```

Comment:
Invalid annotation values should be ignored, instead of errored.

Please output a warning and return nil.